<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Guillermo Cerrillo">

<title>Prácticas - Guillermo Cerrillo - Virulencefinder</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Prácticas - Guillermo Cerrillo</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./virulencefinder.html" aria-current="page"> 
<span class="menu-text">Virulencefinder</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./amrfinder.html"> 
<span class="menu-text">AMRfinder</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./integron.html"> 
<span class="menu-text">Integron Finder</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./orit.html"> 
<span class="menu-text">oriT</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#files" id="toc-files" class="nav-link active" data-scroll-target="#files">Files</a></li>
  <li><a href="#obtaining-the-genome_id" id="toc-obtaining-the-genome_id" class="nav-link" data-scroll-target="#obtaining-the-genome_id">Obtaining the Genome_ID</a></li>
  <li><a href="#getting-rid-of-no-hit-found" id="toc-getting-rid-of-no-hit-found" class="nav-link" data-scroll-target="#getting-rid-of-no-hit-found">Getting rid of “No hit found”</a></li>
  <li><a href="#turning-the-results-into-1" id="toc-turning-the-results-into-1" class="nav-link" data-scroll-target="#turning-the-results-into-1">Turning the results into “1”</a></li>
  <li><a href="#stacking-virulence-classes-in-one-column" id="toc-stacking-virulence-classes-in-one-column" class="nav-link" data-scroll-target="#stacking-virulence-classes-in-one-column">Stacking Virulence classes in one column</a></li>
  <li><a href="#counting-the-occurrence-of-virulences" id="toc-counting-the-occurrence-of-virulences" class="nav-link" data-scroll-target="#counting-the-occurrence-of-virulences">Counting the occurrence of virulences</a></li>
  <li><a href="#plots" id="toc-plots" class="nav-link" data-scroll-target="#plots">Plots</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session Info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Virulencefinder</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Guillermo Cerrillo </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="files" class="level2">
<h2 class="anchored" data-anchor-id="files">Files</h2>
<p>Los archivos usados son los resultados obtenidos con el script <a href="virulencefinder_to_csv.py">virulencefinder_to_csv.py</a> y un compendio de datos sobre todos los genomas disponibles, del cuál solo nos interesa sus datos taxonómicos y su ID.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>data_vf <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"virulencefinder_output.csv"</span>, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>pipolin_summary <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"pipolin_summary_new.tsv"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co">#"viruelencefinder_output.csv" se encuentra comprimdo en un archivo .zip</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>taxonomy_data <span class="ot">&lt;-</span> pipolin_summary[,<span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">34</span>,<span class="dv">38</span><span class="sc">:</span><span class="dv">44</span>)]</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(pipolin_summary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="obtaining-the-genome_id" class="level2">
<h2 class="anchored" data-anchor-id="obtaining-the-genome_id">Obtaining the Genome_ID</h2>
<p>Los datos del Genome_ID se pueden obtener de los hits de cada genoma. Para extraer sólo cierta parte de una string podemos usar la función <code>str_extract()</code> del paquete <code>stringr</code>. En este caso, para todos los genomas que tiene AL MENOS UNA virulencia, se puede extraer de el Genome_ID con sólo extraer la información de las columnas “stx” y “s.aureus_toxin”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># THIS IS TO OBTAIN THE GENOME_ID</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>data_vf_new<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(data_vf[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>)])</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>data_vf_new<span class="sc">$</span>Genome_ID<span class="ot">&lt;-</span><span class="cn">NA</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>data_vf_new<span class="sc">$</span>Genome_ID<span class="ot">&lt;-</span>stringr<span class="sc">::</span><span class="fu">str_extract</span>(<span class="at">string =</span> data_vf_new<span class="sc">$</span>stx, <span class="at">pattern =</span> <span class="st">"G_[0-9]+"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>data_vf_new<span class="sc">$</span>Genome_ID[<span class="fu">which</span>(<span class="fu">grepl</span>(<span class="st">"G_</span><span class="sc">\\</span><span class="st">d+"</span>, data_vf_new<span class="sc">$</span>Genome_ID)<span class="sc">==</span>F)]<span class="ot">&lt;-</span>stringr<span class="sc">::</span><span class="fu">str_extract</span>(<span class="at">string =</span> data_vf_new<span class="sc">$</span>s.aureus_toxin[<span class="fu">which</span>(<span class="fu">grepl</span>(<span class="st">"G_</span><span class="sc">\\</span><span class="st">d+"</span>, data_vf_new<span class="sc">$</span>s.aureus_toxin)<span class="sc">==</span>T)], <span class="at">pattern =</span> <span class="st">"G_[0-9]+"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in data_vf_new$Genome_ID[which(grepl("G_\\d+", data_vf_new$Genome_ID)
== : número de items para para sustituir no es un múltiplo de la longitud del
reemplazo</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>data_vf_new <span class="ot">&lt;-</span> data_vf_new[,<span class="fu">c</span>(<span class="dv">9</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>,<span class="dv">8</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="getting-rid-of-no-hit-found" class="level2">
<h2 class="anchored" data-anchor-id="getting-rid-of-no-hit-found">Getting rid of “No hit found”</h2>
<p>No nos interesa tener información inútil. Así que nos vamos a deshacer de todos los “No hit found” en las 8 columnas de virulencia sustituyéndolos por “NA”. Esto se puede hacer fácil con la función <code>grepl()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This is to get rid of "No hit found"</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>data_vf_binary <span class="ot">&lt;-</span> data_vf_new</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>virulence_ent[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"No hit found"</span>, data_vf_new<span class="sc">$</span>virulence_ent)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="cn">NA</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>listeria[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"No hit found"</span>, data_vf_new<span class="sc">$</span>listeria)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="cn">NA</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>s.aureus_hostimm[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"No hit found"</span>, data_vf_new<span class="sc">$</span>s.aureus_hostimm)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="cn">NA</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>s.aureus_exoenzyme[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"No hit found"</span>, data_vf_new<span class="sc">$</span>s.aureus_exoenzyme)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="cn">NA</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>s.aureus_toxin[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"No hit found"</span>, data_vf_new<span class="sc">$</span>s.aureus_toxin)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="cn">NA</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>virulence_ecoli[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"No hit found"</span>, data_vf_new<span class="sc">$</span>virulence_ecoli)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="cn">NA</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>stx[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"No hit found"</span>, data_vf_new<span class="sc">$</span>stx)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="cn">NA</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>virulence_entfm_entls[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"No hit found"</span>, data_vf_new<span class="sc">$</span>virulence_entfm_entls)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="cn">NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="turning-the-results-into-1" class="level2">
<h2 class="anchored" data-anchor-id="turning-the-results-into-1">Turning the results into “1”</h2>
<p>Como he dicho antes, no necesitamos información inútil en nuestros data frames. Igual que hacíamos antes con los “No hit found”, vamos a cambiar todos los hits por “1”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># THIS IS TO CHANGE RESULTS INTO "1"</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>virulence_ent[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"G_[0-9]+"</span>, data_vf_binary<span class="sc">$</span>virulence_ent)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="dv">1</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>listeria[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"G_[0-9]+"</span>, data_vf_new<span class="sc">$</span>listeria)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="dv">1</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>s.aureus_hostimm[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"G_[0-9]+"</span>, data_vf_new<span class="sc">$</span>s.aureus_hostimm)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="dv">1</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>s.aureus_exoenzyme[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"G_[0-9]+"</span>, data_vf_new<span class="sc">$</span>s.aureus_exoenzyme)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="dv">1</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>s.aureus_toxin[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"G_[0-9]+"</span>, data_vf_new<span class="sc">$</span>s.aureus_toxin)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="dv">1</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>virulence_ecoli[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"G_[0-9]+"</span>, data_vf_new<span class="sc">$</span>virulence_ecoli)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="dv">1</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>stx[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"G_[0-9]+"</span>, data_vf_new<span class="sc">$</span>stx)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="dv">1</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>data_vf_binary<span class="sc">$</span>virulence_entfm_entls[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"G_[0-9]+"</span>, data_vf_new<span class="sc">$</span>virulence_entfm_entls)<span class="sc">==</span>T))]<span class="ot">&lt;-</span><span class="dv">1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="stacking-virulence-classes-in-one-column" class="level2">
<h2 class="anchored" data-anchor-id="stacking-virulence-classes-in-one-column">Stacking Virulence classes in one column</h2>
<p>Como nos interesa analizar los datos según su tipo de virulencia, necesitamos que estas pertenezcan todas a una misma variable. Para ello, nos valemos de la función <code>stack()</code> . También podemos deshacernos de todas las columnas que no presentan virulencia mediante la función <code>filter()</code> del paquete <code>dplyr</code>. También podemos agregar la información taxonómica mediante <code>merge()</code> y deshacernos de columnas duplicadas con <code>unique()</code>.</p>
<p>¿Por qué deshacernos de los duplicados si tenemos una variable Count donde podríamos almacenar la información de cuántas columnas son iguales? Sencillo, porque no nos interesa. Nos basta con saber si esa virulencia está (1) o no está (NA) en nuestro genoma.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>stacked <span class="ot">&lt;-</span> <span class="fu">cbind</span>(data_vf_binary<span class="sc">$</span>Genome_ID,<span class="fu">stack</span>(data_vf_binary[,<span class="dv">2</span><span class="sc">:</span><span class="dv">8</span>]))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(stacked) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Genome_ID"</span>,<span class="st">"Count"</span>,<span class="st">"Virulence"</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>stacked <span class="ot">&lt;-</span> stacked <span class="sc">%&gt;%</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(Count))</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>stacked <span class="ot">&lt;-</span> <span class="fu">unique</span>(stacked)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>data_vf_merge <span class="ot">&lt;-</span> <span class="fu">merge</span>(stacked, taxonomy_data, <span class="at">by=</span><span class="st">"Genome_ID"</span>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>vf_merge_filtered <span class="ot">&lt;-</span> <span class="fu">unique</span>(data_vf_merge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="counting-the-occurrence-of-virulences" class="level2">
<h2 class="anchored" data-anchor-id="counting-the-occurrence-of-virulences">Counting the occurrence of virulences</h2>
<p>Nos interesa saber la cantidad de genomas que tienen 0, 1, 2, 3, etc. virulencias. Para ello hacemos un df con todos los genomas y hacemos un merge con la información taxonómica. Si el merge lo hacemos según Genome_ID nuestro plot no mostrará la cantidad de genomas que no tienen virulencia. Para mostrar también estos últimos, el merge lo haremos para todo el data frame que contiene la información taxonómica.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now to obtain the nº virulences/nº freq plots:</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>Genome_vf_name <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">names</span>(<span class="fu">table</span>(vf_merge_filtered<span class="sc">$</span>Genome_ID)))</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>Genome_vf_count <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">table</span>(vf_merge_filtered<span class="sc">$</span>Genome_ID))</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>df_vf_genome <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(Genome_vf_name, Genome_vf_count)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>df_vf_genome <span class="ot">&lt;-</span> <span class="fu">distinct</span>(df_vf_genome)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(df_vf_genome) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Genome_ID"</span>,<span class="st">"Count"</span>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>df_temp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(vf_merge_filtered<span class="sc">$</span>Genome_ID, vf_merge_filtered<span class="sc">$</span>Genus_rev)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>df_temp <span class="ot">&lt;-</span> <span class="fu">distinct</span>(df_temp)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(df_temp) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Genome_ID"</span>,<span class="st">"Genus_rev"</span>)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>df_vf_genome2 <span class="ot">&lt;-</span> <span class="fu">merge</span>(df_vf_genome, df_temp, <span class="at">by=</span><span class="st">"Genome_ID"</span>)</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>df_vf_genome3 <span class="ot">&lt;-</span> <span class="fu">merge</span>(taxonomy_data, df_vf_genome2, <span class="at">all.x =</span> T)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>df_vf_genome3<span class="sc">$</span>Count[<span class="fu">which</span>((<span class="fu">grepl</span>(<span class="st">"[0-9]+"</span>, df_vf_genome3<span class="sc">$</span>Count)<span class="sc">==</span>F))]<span class="ot">&lt;-</span><span class="dv">0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plots" class="level2">
<h2 class="anchored" data-anchor-id="plots">Plots</h2>
<p>El primer gráfico muestra la frecuencia relativa de presencia de cierta viruelncia, coloreada según pertenencia a los géneros más relevantes de bacterias y dividido por clase taxonómica.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(vf_merge_filtered, <span class="fu">aes</span>(<span class="at">x=</span>Virulence))<span class="sc">+</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(count)<span class="sc">/</span><span class="dv">11430</span>, <span class="at">fill =</span> Genus_rev, <span class="at">alpha=</span><span class="fl">0.1</span>), <span class="at">stat =</span> <span class="st">"count"</span>, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">alpha =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(.<span class="sc">~</span>class, <span class="at">scale=</span><span class="st">"free_x"</span>,<span class="at">space=</span><span class="st">"free"</span>) <span class="sc">+</span> <span class="fu">scale_y_continuous</span>(<span class="at">labels=</span>scales<span class="sc">::</span>percent) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"VIRULENCE"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Relative Freq"</span>) <span class="sc">+</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="dv">10</span>, <span class="at">face =</span> <span class="st">"bold"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="virulencefinder_files/figure-html/plot1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>El segundo muestra qué cantidad de genomas tienen x número de virulencias, sólo para genomas que tengan al menos una.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_vf_genome2, <span class="fu">aes</span>(Count, <span class="at">fill=</span>Genus_rev)) <span class="sc">+</span> <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"count"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"Nº VIRULENCES"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Nº GENOMES"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="virulencefinder_files/figure-html/plot2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>El tercero muestra qué cantidad de genomas tienen x número de virulencias, incluyendo aquellos que no presentan ninguna.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_vf_genome3, <span class="fu">aes</span>(Count, <span class="at">fill=</span>Genus_rev)) <span class="sc">+</span> <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"count"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"Nº VIRULENCES"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Nº GENOMES"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="virulencefinder_files/figure-html/plot3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="session-info" class="level2">
<h2 class="anchored" data-anchor-id="session-info">Session Info</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 11 x64 (build 22631)

Matrix products: default


locale:
[1] LC_COLLATE=Spanish_Spain.utf8  LC_CTYPE=Spanish_Spain.utf8   
[3] LC_MONETARY=Spanish_Spain.utf8 LC_NUMERIC=C                  
[5] LC_TIME=Spanish_Spain.utf8    

time zone: Europe/Madrid
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.2     readr_2.1.4     tidyr_1.3.0     tibble_3.2.1   
 [9] tidyverse_2.0.0 ggplot2_3.4.4  

loaded via a namespace (and not attached):
 [1] bit_4.0.5         gtable_0.3.4      jsonlite_1.8.8    crayon_1.5.2     
 [5] compiler_4.3.2    tidyselect_1.2.0  parallel_4.3.2    scales_1.3.0     
 [9] yaml_2.3.7        fastmap_1.1.1     R6_2.5.1          labeling_0.4.3   
[13] generics_0.1.3    knitr_1.45        htmlwidgets_1.6.3 munsell_0.5.0    
[17] pillar_1.9.0      tzdb_0.4.0        rlang_1.1.2       utf8_1.2.4       
[21] stringi_1.8.2     xfun_0.41         bit64_4.0.5       timechange_0.2.0 
[25] cli_3.6.1         withr_2.5.2       magrittr_2.0.3    digest_0.6.33    
[29] grid_4.3.2        vroom_1.6.4       rstudioapi_0.15.0 hms_1.1.3        
[33] lifecycle_1.0.4   vctrs_0.6.4       evaluate_0.23     glue_1.6.2       
[37] farver_2.1.1      fansi_1.0.5       colorspace_2.1-0  rmarkdown_2.25   
[41] tools_4.3.2       pkgconfig_2.0.3   htmltools_0.5.7  </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>